mypy/argmap.py

```diff
@@ -265,3 +276,189 @@ def expand_actual_type(
+
+        # We first create an upcast function:
+        #    def [T] (Iterable[T]) -> Iterable[T]: ...
+        # and then solve for T, given the input type as the argument.
+        T = TypeVarType(
+            "T",
+            "T",
+            TypeVarId(-1),
+            values=[],
+            upper_bound=AnyType(TypeOfAny.special_form),
+            default=AnyType(TypeOfAny.special_form),
+        )
+        target = Instance(iterable_kind, [T])
+
+        upcast_callable = CallableType(
+            variables=[T],
+            arg_types=[target],
+            arg_kinds=[ARG_POS],
+            arg_names=[None],
+            ret_type=T,
+            fallback=self.context.function_type,
+        )
+        constraints = infer_constraints_for_callable(
+            upcast_callable, [typ], [ARG_POS], [None], [[0]], context=self.context
+        )
+
+        (sol,), _ = solve_constraints([T], constraints)
+
+        if sol is None:  # solving failed, return AnyType fallback
+            return AnyType(TypeOfAny.from_error)
+        return self._make_iterable_instance_type(sol)
+
+    def as_iterable_type(self, typ: Type) -> IterableType | AnyType:
+        """Reinterpret a type as Iterable[T], or return AnyType if not possible."""
+        p_t = get_proper_type(typ)
+        if self.is_iterable_instance_type(p_t) or isinstance(p_t, AnyType):
+            return p_t
+        elif isinstance(p_t, UnionType):
+            # If the type is a union, map each item to the iterable supertype.
+            # the return the combined iterable type Iterable[A] | Iterable[B] -> Iterable[A | B]
+            converted_types = [self.as_iterable_type(get_proper_type(item)) for item in p_t.items]
+
+            if any(not self.is_iterable_instance_type(it) for it in converted_types):
+                # if any item could not be interpreted as Iterable[T], we return AnyType
+                return AnyType(TypeOfAny.from_error)
+            else:
+                # all items are iterable, return Iterable[T₁ | T₂ | ... | Tₙ]
+                iterable_types = cast(list[IterableType], converted_types)
+                arg = make_simplified_union([it.args[0] for it in iterable_types])
+                return self._make_iterable_instance_type(arg)
+        elif isinstance(p_t, TupleType):
+            # maps tuple[A, B, C] -> Iterable[A | B | C]
+            # note: proper_elements may contain UnpackType, for instance with
+            #   tuple[None, *tuple[None, ...]]..
+            proper_elements = [get_proper_type(t) for t in flatten_nested_tuples(p_t.items)]
+            args: list[Type] = []
+            for p_e in proper_elements:
+                if isinstance(p_e, UnpackType):
+                    r = self.as_iterable_type(p_e)
+                    if self.is_iterable_instance_type(r):
+                        args.append(r.args[0])
+                    else:
+                        # this *should* never happen
```

Comment:
`assert` then?

Reply:
I'd transform this to `assert self.is_iterable_instance_type(r);  args.append(r.args[0])`, but depends on how certain you are. If there are some unlikely cases where this happens and does not indicate a programming error, your current solution is also fine